### DISCUSSION TITLE
Client-Side shell call

### DISCUSSION THREAD

**Discussion by trixtan** [eclipse-theia/theia#11978] [Q&A] [answered]
Hello,

I am trying to write an IDE for C/C++ based embedded development based on Theia. The IDE will be cloud based, but 
embedded development requires certain components to be run on the client side, where also the Hardware is available.
A typical workflow would be:

- Write a program using the Editor
- (Server side) - Build server side using a cross-compiler -> this generates an HEX file
- (Client side) Download and Flash the generated HEX file
- (Client side) Debug, trace, etc.

Is this scenario feasible using the Theia framework? What I basically need is a way to start a process on the client side (flasher, Debugger Adapter) in a transparent way. The user shall not be required to manually download and install a flasher, for example, but the IDE shall be able to "inject" all required dependencies to be run on the client.

Thank you,
Best regards
Enrico

**Comment by msujew** [comment]
Hey @trixtan,

I'm not 100% sure, but I believe that ARM does something like that with their [mbed studio](https://os.mbed.com/studio/). I believe they're using the [USB browser API](https://developer.mozilla.org/en-US/docs/Web/API/USB) to talk with the hardware from the JS frontend code. Maybe @thegecko can tell you more about this.

**Reply by trixtan** [reply]
Hello @msujew . Thank you for the hint. Just for me to be sure, are you referring to the [WebUSB API](https://en.wikipedia.org/wiki/WebUSB) ?
Thank you, kind regards

**Reply by msujew** [reply]
I think that should be the one they're using. ARM recently contributed a fix to Theia to allow access to that API within webviews, see https://github.com/eclipse-theia/theia/pull/10848.

**Reply by trixtan** [reply]
Thank you!

**Comment by thegecko** [comment, marked as answer]
[Mbed Studio](https://os.mbed.com/studio/) is a desktop application using the Theia framework, our web application is [Keil Studio](https://studio.keil.arm.com/) which uses the same codebase.

On the desktop app you can call whatever you want on the client and we use USB access directly to interact with devices.

In the browser you are restricted to technologies such as [WebUSB](https://wicg.github.io/webusb/) which we use and its not possible to invoke local processes without some sort of agent.

Arm have released a few VS Code extensions for interacting with embedded devices, flashing and debugging them which work with VS Code and Theia apps on the desktop and on the web (using WebUSB):

- https://open-vsx.org/extension/arm/device-manager
- https://open-vsx.org/extension/arm/embedded-debug

Depending on the hardware you are trying to use, these could do what you want already.

**Reply by trixtan** [reply]
Hi @thegecko , thank you for the interesting summary and the links.

I gave a look to the WebUSB API, and from what I understand I would have to rewrite my tools to run in the browser, am I right? Let's take the flasher, for example. I am working with the Infineon AURIX platform, which comes with its own AURIXFlasher.exe and leverages the proprietary MCD API. I would need to rewrite the flasher logic AND the MCD API in TypeScript, right? (They are both native components written in C#/C++) .

Talking about the other approach you mentioned (the local agent): do you mean a local, native process which is able to interact with the Theia UI via a TCP socket or something similar?

Thank you for the clarifications,
kind regards
Enrico

**Comment by thegecko** [comment]
Yes on both counts

**Reply by trixtan** [reply]
Thank you.
